### ISSUE TITLE
Crash when loading "µcomputer" patch on Japanese Windows 10

### ISSUE THREAD

**Issue by mvf** [surge-synthesizer/surge#2579]
**TL;DR:** Uncaught C++ exception because Surge doesn't support Unicode paths on Windows.

Related to #1699

**Description**
Trying to load the "Lead"→"µcomputer" patch on a Japanese Windows 10 crashes Surge.

* Surge Version: 1.7.main.502ec53a
* Plugin type: VST3
* Bits: 64

**To Reproduce**
1. Make sure your ANSI code page does not have the "µ" character
2. In the patch browser menu, click "Lead"→"µcomputer"
3. Observed: Crash

**Expected behavior**
The "µcomputer" patch gets loaded.

**Desktop:**
 - OS: Windows 10 Professional 64-Bit, Japanese
 - Host: VST3PluginTestHost from the VST3 SDK
 - Version: 2.9.10 Build 39

**Additional context**
The immediate cause of the crash is an uncaught `std::system_error` from `std::filesystem::path::generic_string()` called here:

```c++
void SurgeSynthesizer::loadPatch(int id)
{
   ...
   loadPatchByPath( e.path.generic_string().c_str(), e.category, e.name.c_str() ); // generic_string() throws
}
```

The exception is thrown because code page 932 on a Japanese Windows doesn't have the "µ" character. Screenshot of a debug session at the relevant `WideCharToMultiByte()` call:

![surge-cropped](https://user-images.githubusercontent.com/304760/92328353-8d843580-f060-11ea-84ef-d6f821ef4835.png)
`_Code_page` is `0`, aka `CP_ACP`, i.e. the system default ANSI code page. `_Used_default_char` being `1` means that not all characters were representable in this code page. ([WideCharToMultiByte Docs](https://docs.microsoft.com/en-us/windows/win32/api/stringapiset/nf-stringapiset-widechartomultibyte))

`std::filesystem::path::generic_string()` seems to be used in multiple places, so this might be a broader issue. On Western versions of Windows, this should be reproducible with paths containing CJK characters.

**Comment by baconpaul** [comment]
Thank you for the detailed report

Do you have any suggestions on a solution here? I have basically no experience with windows Unicode programming.

Of course we could wrap that in a try catch and show a user error but that seems less than ideal. The internal load apis are all const char* based and ultimately call Fopen.

**Comment by mvf** [comment]
Agreed, an error message doesn't seem like a great user experience. There might even be cases where a message can't be displayed or where the failure isn't clearly connected to a specific user action.

Basically, to support Unicode on Windows, Surge...

- ...must not call `std::filesystem::path::generic_string()`
- ...needs to talk UTF-16 to the filesystem

I see two ways to achieve this:

#### Approach A) Let paths be `std::filesystem::path`s

Get rid of `fopen()` and instead construct `fstream`s directly from `path` objects. This should make the problematic conversion to `string` all but unnecessary. It also communicates intent better throughout the codebase, and the RAII-ness of `fstream`s prevents leaks.

If there's code that *really* needs to process paths as strings, it should be implemented in terms of `path`'s `string_type` and `value_type` to avoid encoding conversions. Alternatively, explicit UTF-8 is also available via `generic_u8string()`/`u8path()`.

#### Approach B) Emulate UTF-8 on Windows

If A) is too much work, path-ey `string`s can be made UTF-8 and converted to UTF-16 at the Windows API boundary, i.e.:

1. Add a wrapper function that takes a `path` and returns `path.generic_u8string()` on Windows and `path.generic_string()` everywhere else.
2. Replace all other `generic_string()` calls with a call to this function.
3. On Windows, link an `fopen()` implementation that converts its arguments to UTF-16 and passes them to [`_wfopen()`](https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/fopen-wfopen?view=vs-2019).
4. Repeat step 3 for all other library functions that involve paths. The vanilla ANSI functions will now only be able to access 7-bit ASCII paths, i.e. the intersection of UTF-8 and the ANSI code page.

**Comment by baconpaul** [comment]
Oof.

So a is hard because, In order to support macos 10.11 we have to implement filesystem ourselves on Mac (it isn’t available until the 10.12 api) and the rack compilation paths also don’t have std::fs consistently available. And older linuxes have different names. We deal with this at cmake time and we do have a stub hand coded library but we don’t push the api all the way down the codebase. It would be a lot of surgery. 

And b is hard because we would of course miss somewhere I’m sure. The xml parser uses fopen internally for instance but that is probably less of an issue since we have somewhat better control over those names. We could at first blush just address the patch loading. 

I’ll need to think about this.

You seem very well versed in this topic. Would you have time or inclination to write some of this code?

**Comment by mkruselj** [comment]
Yeah your description of the issue and suggested solutions give off a flair of you really knowing this stuff. We are in a dire need of a good Windows-first programmer, so you are very welcome to join! Hit us up on our Discord channel if you're interested!

**Comment by baconpaul** [comment]
Looking at the code, there's a lot of fopen calls only used in tests and examples or on linux and mac but the ones which would absolutely need resolution are these

```
src/common/SurgeStorage.cpp:   FILE* f = fopen(filename.c_str(), "rb");
src/common/WavSupport.cpp:    FILE *fp = fopen(fn.c_str(), "rb");
src/common/WavSupport.cpp:      FILE *wfp = fopen( fname.c_str(), "wb" );
src/common/SurgeSynthesizerIO.cpp:   FILE* f = fopen(fxpPath, "rb");
src/windows/UserInteractionsWin.cpp:    FILE *f = fopen(fns.str().c_str(), "w" );
libs//xml/tinyxml.cpp:  FILE* file = fopen( value.c_str (), "r" );
libs//xml/tinyxml.cpp:  FILE* fp = fopen( filename, "w" );
libs//nanosvg/src/nanosvg.h:    fp = fopen(filename, "rb");
libs//nanosvg/src/nanosvgrast.h:        FILE* fp = fopen(name, "w");
```

Changing WavSupport from fopen/fread to streams is a bit painful. And nanosvg is a submodule so we would have to branch it or just never use non-ascii characters in svgs file names. The UserInteractionWin is just dumping a tmp html file so is easy to swap to streams but would fail if `GetTempPath` returned an invalid character.

Funnily, `savePatch` in SurgeSynthesizerIO.cpp already uses fs::path so (outside of VCV Rack) the save side is fine. (And the save side doesn't trigger in rack).

**Comment by mkruselj** [comment]
I think we should be good with restricting SVG filenames to non-ascii (should be documented in skin engine), so at least one pain less.

**Comment by mvf** [comment]
> Would you have time or inclination to write some of this code?

I'll try to cook something up. If all goes to plan, B) shouldn't be too much code.

> a is hard

Agreed. It'd be quite some effort. I also wasn't aware that not all target platforms are fully C++17, so this makes no sense ATM.

> And b is hard because we would of course miss somewhere I’m sure.

Those nasty ANSI paths can't come from too many places. We should be able to Catch 'Em All™.

> you are very welcome to join!

Sure, I'd be happy to help out as time permits. I'll get in touch!

**Comment by baconpaul** [comment]
Yeah indeed; in fact the only real one you need to worry about is the fopen for patch I think, as I mentioned above. 

Would be happy to work with you on this! We are really active on discord too if you want something more realtime.

Thank you for the help, even in just identifying the problem. It's appreciated.
